### ISSUE TITLE
useChain does not work as expected for transitions

### ISSUE THREAD

**Issue by martinkutter** [pmndrs/react-spring#1453]
## 🐛 Bug Report

`useChain` does not chain animations correct, if `Transition` should animate after `Spring`.

## To Reproduce
`Spring` first, than `Transition` -> does not work
See https://codesandbox.io/s/wizardly-leaf-b4v5g?file=/src/App.tsx

`Transition` first, than `Spring` -> does work
comment out line 38; comment in line 40
## Expected behavior
`Transition` animation should run after `Spring` animation, not in sequence.

## Environment

- `react-spring` v9.1.2
- `react` v17.0.1

**Comment by martinkutter** [comment]
What I have found out so far: 
I think the issue is, that there are no controllers available for transitions in https://github.com/pmndrs/react-spring/blob/master/packages/core/src/hooks/useChain.ts#L15

**Comment by joshuaellis** [comment]
Firstly, I think you need to set an actual height on the inner `divs` the don't appear and it's confusing.
Secondly, and this is more so i understand, you've removed the data from the transition, but you're expecting the divs to stay there until the spring has changed background color and then the inner divs animate out?

**Comment by martinkutter** [comment]
Ups... :( I had forgotten to save the CSS file.
I think the Example is now much clearer.

https://codesandbox.io/s/wizardly-leaf-b4v5g?file=/src/App.tsx

https://user-images.githubusercontent.com/16589094/116391969-24844480-a820-11eb-84de-b9d48906838d.mp4
In the first part of the screen capture the animation starts parallel not chained. In the second part the chain starts with the transition and the animations runs chained as expected.

**Comment by joshuaellis** [comment]
So the issue is the background color changes whilst the transitions are animating in. but actually, the background color should change **then** the divs animate in?

**Comment by martinkutter** [comment]
Yes correct!

**Comment by martinkutter** [comment]
I had the chance to take a closer look at the problem with the missing controller.

If I understand correct at this line the `controller` should be assigned to `ref` from `props`.
https://github.com/pmndrs/react-spring/blob/9568020efb64addb97816232759528a7681ed97c/packages/core/src/hooks/useTransition.tsx#L295

But `payload.ref` is not (and cannot be) defined. `payload` is initialized in 
https://github.com/pmndrs/react-spring/blob/9568020efb64addb97816232759528a7681ed97c/packages/core/src/hooks/useTransition.tsx#L206
`defaultProps` come from https://github.com/pmndrs/react-spring/blob/9568020efb64addb97816232759528a7681ed97c/packages/core/src/hooks/useTransition.tsx#L168
But `ref` is not in `DEFAULT_PROPS` https://github.com/pmndrs/react-spring/blob/9568020efb64addb97816232759528a7681ed97c/packages/core/src/helpers.ts#L96-L104

**Comment by joshuaellis** [comment]
It sounds like you're gonna have a fix quicker than I, feel free to make a PR for this ⭐ although I will be looking for us to include tests to ensure refs that are passed are assigned and working.

**Comment by joshuaellis** [comment]
this has been released in `9.2.0-beta.1`

**Comment by martinkutter** [comment]
Unfortunately my changes breaks some transitions completely if a `ref` is injected.
Compare 9.2.0-beta.0 and 9.2.0-beta.1 in https://codesandbox.io/s/flamboyant-violet-7nbv3?file=/src/App.tsx

I try to figure out why for some hours...
Now that a `ref` may exists there is something needed to adjust here:
https://github.com/pmndrs/react-spring/blob/922f6407df005f57880d5257e9889134ad6cd6a8/packages/core/src/hooks/useTransition.tsx#L302-L303

Unfortunately I have no deeper insight here, maybe you can help?

**Comment by joshuaellis** [comment]
Okay I will revert your commit because we can't have a regression.

Regarding me looking at it, I will assign the ticket to myself to look at when I have time too.

**Comment by martinkutter** [comment]
Maybe a little support, it initially breaks with the `SpringRef` [rework](https://github.com/pmndrs/react-spring/commit/92348532a0557b81608504cbdd36da3a015567b8).

**Comment by martinkutter** [comment]
Hey @joshuaellis, I imagine you have a lot of work to do. Nevertheless, because my current project relies very heavily on chained animations, I would like to ask carefully if there is a forecast when you could look at this bug? Or maybe there is another contributor who could at least give me a hint where the problem should be solved? Thank you! 🙏

**Comment by joshuaellis** [comment]
Hi @martinkutter, unfortunately I can't tell you when i'll be able to work on this, if you're interested in enabling me to put more time into the project, feel free to sponsor me so I don't have to do as much client work. But it is on my radar for this weekend time permitting.

**Comment by joshuaellis** [comment]
> Unfortunately my changes breaks some transitions completely if a `ref` is injected.
> Compare 9.2.0-beta.0 and 9.2.0-beta.1 in https://codesandbox.io/s/flamboyant-violet-7nbv3?file=/src/App.tsx
> 
> I try to figure out why for some hours...
> Now that a `ref` may exists there is something needed to adjust here:
> https://github.com/pmndrs/react-spring/blob/922f6407df005f57880d5257e9889134ad6cd6a8/packages/core/src/hooks/useTransition.tsx#L302-L303
> 
> Unfortunately I have no deeper insight here, maybe you can help?

In this example you're trying to pass a regular React Ref, which is no longer possible across the entire library. Was this the only issue? Or did I miss something. Because if it is, I think you did solve the issue...

**Comment by martinkutter** [comment]
Thanks for your time! Oh... that was something I changed while testing several versions. I updated the codesandbox and use `SpringRef` again. But as you can see, this simple transition works in 9.2.0-beta.0 but breaks in 9.2.0-beta.1.

By the way, sponsoring is a great idea. I will talk to my project owner next week.

**Comment by joshuaellis** [comment]
SO, I've had a look at it, yeah this line:

```
ctrl[ctrl.ref ? 'update' : 'start'](payload) 
```

is important, it's what keeps our API consistent, i.e. if you ever pass a ref, you must call `ref.start` to start the transition, here's a forked sandbox demonstrating this: https://codesandbox.io/s/goofy-allen-u4hml?file=/src/App.tsx

**Comment by martinkutter** [comment]
Thank you very much for your time and your PR. While for me it is a little confusing that I have to call `start()` in this case, but not if I use `useChain` it's good to hear that it is only a docs and understanding issue. I will check if everything works fine in our project in the next days.

**Comment by joshuaellis** [comment]
Well `useChain` is designed to trigger animations, consistent with other hooks when you pass a ref, so imo it's better this way, i'll be releasing a `beta` once another PR is ready.

**Comment by joshuaellis** [comment]
This has been released on v9.2.0-beta.4
